lib/common/ipc_client.c

```diff
@@ -1361,44 +1376,18 @@ crm_ipc_send(crm_ipc_t *client, const xmlNode *message,
         goto send_cleanup;
     }
 
-    /* Read the IPC response, respecting any timeout we were passed */
-    if (ms_timeout > 0) {
-        rc = internal_ipc_get_reply(client, header->qb.id, ms_timeout, &bytes);
-        if (rc != pcmk_rc_ok) {
-            /* We didn't get the reply in time, so disable future sends for now.
-             * The only alternative would be to close the connection since we
-             * don't know how to detect and discard out-of-sequence replies.
-             *
-             * @TODO Implement out-of-sequence detection
-             */
-            client->need_reply = TRUE;
-        }
-        rc = (int) bytes; // Negative system errno, or size of reply received
-
-    } else {
-        do {
-            qb_rc = qb_ipcc_recv(client->ipc, client->buffer, client->buf_size,
-                                 -1);
-        } while ((qb_rc == -EAGAIN) && crm_ipc_connected(client));
-        rc = (int) qb_rc; // Negative system errno, or size of reply received
+    rc = internal_ipc_get_reply(client, header->qb.id, ms_timeout, &bytes, reply);
+    if (rc != pcmk_rc_ok) {
+        /* We didn't get the reply in time, so disable future sends for now.
+         * The only alternative would be to close the connection since we
+         * don't know how to detect and discard out-of-sequence replies.
+         *
+         * @TODO Implement out-of-sequence detection
+         */
+        client->need_reply = TRUE;
```

Comment:
Previously, we were setting this on failure only when a timeout is enabled. What is the rationale for setting it across-the-board now? I haven't traced through the relevant code, I'm about done for the night, hoping you can tell me.

Reply:
I think we probably should still be setting it only when there's a timeout.  Fixed.

Reply:
Looks like there's not much tracing to do after all :) From a skim, it looks like the point of it is to wait for an old reply so that we can discard it. That **might** be necessary upon error even without a timeout... but I think you're right